src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,966 @@
+            #red(mapRec l, x, f(x,y), mapRec r)
+          };
+          case (#black(l, x, y, r)) {
+            #black(mapRec l, x, f(x, y), mapRec r)
+          };
+        }
+      };
+      mapRec(rbMap)
+    };
+
+    public func foldLeft<Key, Value, Accum>(
+      rbMap : Map<Key, Value>,
+      base : Accum,
+      combine : (Key, Value, Accum) -> Accum
+    ) : Accum
+    {
+      switch (rbMap) {
+        case (#leaf) { base };
+        case (#red(l, k, v, r)) {
+          let left = foldLeft(l, base, combine);
+          let middle = combine(k, v, left);
+          foldLeft(r, middle, combine)
+        };
+        case (#black(l, k, v, r)) {
+          let left = foldLeft(l, base, combine);
+          let middle = combine(k, v, left);
+          foldLeft(r, middle, combine)
+        }
+      }
+    };
+
+    public func foldRight<Key, Value, Accum>(
+      rbMap : Map<Key, Value>,
+      base : Accum,
+      combine : (Key, Value, Accum) -> Accum
+    ) : Accum
+    {
+      switch (rbMap) {
+        case (#leaf) { base };
+        case (#red(l, k, v, r)) {
+          let right = foldRight(r, base, combine);
+          let middle = combine(k, v, right);
+          foldRight(l, middle, combine)
+        };
+        case (#black(l, k, v, r)) {
+          let right = foldRight(r, base, combine);
+          let middle = combine(k, v, right);
+          foldRight(l, middle, combine)
+        }
+      }
+    };
+
+    public func mapFilter<K, V1, V2>(rbMap : Map<K, V1>, compare : (K, K) -> O.Order, f : (K, V1) -> ?V2) : Map<K, V2>{
+      func combine(key : K, value1 : V1, acc : Map<K, V2>) : Map<K, V2> {
+        switch (f(key, value1)){
+          case null { acc };
+          case (?value2) {
+            put(acc, compare, key, value2)
+          }
+        }
+      };
+      foldLeft(rbMap, #leaf, combine)
+    };
+
+    public func size<K, V>(t : Map<K, V>) : Nat {
+      switch t {
+        case (#red(l, _, _, r)) {
+          size(l) + size(r) + 1
+        };
+        case (#black(l, _, _, r)) {
+          size(l) + size(r) + 1
+        };
+        case (#leaf) { 0 }
+      }
+    };
+
+    public func get<K, V>(t : Map<K, V>, compare : (K, K) -> O.Order, x : K) : ?V {
+      switch t {
+        case (#red(l, x1, y1, r)) {
+          switch (compare(x, x1)) {
+            case (#less) { get(l, compare, x) };
+            case (#equal) { ?y1 };
+            case (#greater) { get(r, compare, x) }
+          }
+        };
+        case (#black(l, x1, y1, r)) {
+          switch (compare(x, x1)) {
+            case (#less) { get(l, compare, x) };
+            case (#equal) { ?y1 };
+            case (#greater) { get(r, compare, x) }
+          }
+        };
+        case (#leaf) { null }
+      }
+    };
+
+    func redden<K, V>(t : Map<K, V>) : Map<K, V> {
+      switch t {
+        case (#black (l, x, y, r)) {
+          (#red (l, x, y, r))
+        };
+        case _ {
+          Debug.trap "RBTree.red"
```

Comment:
```suggestion
          Debug.trap "PersistentOrderedTree.red"
```
Are there others that need fixing?

Reply:
Fixed, I haven't found more "RBTree" stuff in the code.